proxy.go

```diff
@@ -548,6 +548,17 @@ func (proxy *proxy) serveNewClient(proto *protocol, newConn net.Conn) {
 		newClient.infof(1, "error serving client: %v", err)
 	}
 
+	// The client was a shim with a session still alive. This means DisconnectShim
+	// hasn't been called and we're closing the connection because of an error
+	// (either us or the shim has closed the connection). We want to keep the
+	// session alive and hope a shim will reconnect claiming the same token.
+	if newClient.session != nil {
+		proxy.Lock()
+		info := proxy.tokenToVM[newClient.token]
+		info.state = tokenStateAllocated
```

Comment:
Is this session being kept round forever. If the shim disconnects, we should wait for a reconnection for a specific time, after which we should discard the session.